### ISSUE TITLE
Support running backups in parallel

### ISSUE THREAD

**Issue by watn3y** [garethgeorge/backrest#221]
**Is your feature request related to a problem? Please describe.**
Backrest doesn't seem to support running multiple backups simultaneously. Since restic itself supports running `backup` without exclusively locking the repo, I see no reason why backrest wouldn't.

**Describe the solution you'd like**
A config option that limits parallel backups either globally or on a per-repo basis.

**Comment by garethgeorge** [comment]
Hmm, thinking through this I can see that the idea of parallel backup options is immediately appealing as it sounds like it should be a speedup -- but I'm interested to think through what the real performance gains will be (and what costs / risks if any may come with concurrent backups):

Things that jump out to me are 

 * Parallel backups aren't actually a speedup if you are network IO limited e.g. by your upload speed.
 * Parallel backups aren't actually a speedup if you are disk IO limited e.g. all of the repos you are scanning are on a single HDD.
 * Parallel backups aren't always safe with a concurrent prune operation (though restic locks will typically block this case). If parallelism is added it should be single-concurrency on a given target repository, parallelism across repos.

I see parallelism as introducing some risks and backrest does aim to be opinionated in places where it can avoid "footguns" e.g. places where a user can accidentally break themself or simply may not stand to gain.

I'm curious how much value it adds in the places you may be using backrest / how much speedup you're expecting?

**Comment by watn3y** [comment]
You make some valid points, but I still believe that in some cases it would be beneficial to run backups in parallel. 

For example:
I am running backrest on a Server with a 10G Uplink. I have it connected to 2 repos, each with a 1G Link. 

Every 15 minutes I run a backup of some smaller files to repo 1.
Once a day I back up my larger files to repo 2, this usually takes up to an hour.

While the daily backup to repo 2 is running, backrest doesn't start any of the 4 scheduled backups to repo 1.

An option to run backups to different repos independently of each other would be great to have here.

**Comment by brandonkal** [comment]
Parallel backups generally will be a speedup. I am running a backup and the bottleneck is the bandwidth+latency of the S3 service, not my upload speed.

Additionally, it would be nice to be able to pause a backup task.

**Comment by garethgeorge** [comment]
OK, I'm open to this but I think I'm going to consider it low priority for now -- I think my near term focus is continuing to improve, reliability, unblock some workflows by improving hook handling, and add multi-host management as a feature. I'll defer parallel execution for now as it makes some of that more challenging. Looking forward though, architecturally Backrest  does have the right concurrency controls in place to make this possible both on the backend and on the networking side.

**Comment by watn3y** [comment]
I'm with you on that. Thanks for considering it, and thanks for this great piece of software :)

**Comment by umrath** [comment]
I was quite astonished when I realized that the (very slow) backup from the cifs mounted Homeassistant is blocking all other jobs and couldn't find a way to enable parallel backups (which restic itself is just fine with).

I do understand your concerns.
But there really are quite a number of situations where parallel backups will significantly speed up the process.

**Comment by garethgeorge** [comment]
I think I'm reasonably comfortable moving this up the priority list; Backrest can move to a model where operations can run in parallel across repos (but are still serialized within a repo).

**Comment by umrath** [comment]
This would be a huge step forward.

In preparation, I have already created a dedicated repository for every job.
I'm not sure whether this is what restic was intended to be used like. But if helps to run backups in parallel, then I'm fine with it.

The ultimate solution would still be to implement what restic can do: start multiple backup jobs in parallel on a single repository (without a global lock on the repository.

I'm unsure whether the multiple repository strategy has unwanted side effects (e.g. lower compression, higher overhead, etc.)
But that's probably absolutely minor compared to the benefit of being able to run parallel backups.

**Comment by umrath** [comment]
I just realized that this still has some side effects:
Backing up to dedicated repositories will indeed result in a much higher disk usage. I do have plenty of duplicates on different systems that I backup. reading https://forum.restic.net/t/how-does-restic-deduplicate/2761 I would expect restic to be able to consolidate this and safe a ton of space - but only if the data resides in one repository.
If I (have to) create dedicated repositories for every backup job to be able to get parallel job execution, this will nullify the functionality.

**Comment by thierrybla** [comment]
Hi, is there any update on this? It's slowing down my updates by a lot.

**Comment by nsubtil** [comment]
This has already been mentioned in this issue, but adding my +1 here as motivation: not being able to run concurrently means that unrelated backups can get delayed significantly. My use case involves backups that occasionally need to update hundreds of GB in one go, and these end up stalling other backups that are on the order of dozens of MB and use different repositories.

Remote storage is quite slow in my case, hundreds of GB can take a few of days to upload, but scales very well with concurrency.

**Comment by nadres** [comment]
+1 as well. My use case is for my always-on server, daily backing up two different sets of data on different HDDs, each to one repository located in different external HDDs.

Dataset 1 in SATA3 HDD1 -> Repo A in USB3 HDD1.
Dataset 2 in SATA3 HDD2 -> Repo B in USB3 HDD2.

I would love to have that done in parallel.

As an aside, I started using backrest/restic for a week now, coming from made up rsync scripting for back up and feels like I leveled up my backup strategy! Thanks much for this piece of software.

**Comment by umrath** [comment]
This issue is getting more pressing for me now.
I have a remote Nextcloud instance with lots of data. The initial backup takes several weeks, because access is rather slow.

It also means that my regular backups are not running at all during that time, even though performance or integrity would really not be an issue. At all.

Any projection on when this might end up on the roadmap?

**Comment by mcbrineellis** [comment]
Hi there Gareth, I just discovered this project and it has been TERRIFIC so far 🖤😄!

This is the first issue I noticed in my testing, a very slow and long-running backup to some cloud storage is preventing any other jobs from running even if they are to a different repo.  So I would greatly appreciate having the option to turn on parallel operations.

**Comment by dullage** [comment]
+1 on this.

Backrest is great and made me switch from Duplicati to restic. But I do have a use case where 1 of the 3 plans on one machine can occasionally take a day or two to complete a backup, which holds up the other 2 plans. 

> Backrest can move to a model where operations can run in parallel across repos (but are still serialized for plans backing up to the same repo).

In my case, all 3 plans point to the same repo, so the above solution would still leave me with a problem. For me, it's not about running parallel to speed things up, it's more about letting the smaller backups complete whilst a bigger backup is still running.

My current workaround is to run 2 instances of Backrest on the same machine (pointing to the same repo). It'd be great if this wasn't needed.

Thanks for all your hard work.

**Comment by IetIesAai** [comment]
just throwing in mo 2 cents - I also would like some parralellism, mainly so large jobs (eg media files) do not throw off the schedule for smaller jobs (eg configfiles)
